File: plugins/plugin-elizacloud/rust/src/actions/check_credits.rs
//! CHECK_CLOUD_CREDITS — Query ElizaCloud credit balance and usage.

use std::collections::HashMap;

use crate::cloud_api::CloudApiClient;
use crate::cloud_types::{ActionResult, ContainerStatus};
use crate::error::Result;
use crate::services::CloudContainerService;

pub const ACTION_NAME: &str = "CHECK_CLOUD_CREDITS";
pub const ACTION_DESCRIPTION: &str =
    "Check ElizaCloud credit balance, container costs, and estimated remaining runtime.";

const DAILY_COST_PER_CONTAINER: f64 = 0.67;

/// Handle the CHECK_CLOUD_CREDITS action.
pub async fn handle_check_credits(
    client: &CloudApiClient,
    container_svc: Option<&CloudContainerService>,
    options: &HashMap<String, serde_json::Value>,
) -> Result<ActionResult> {
    let resp = client.get("/credits/balance").await?;
    let balance = resp
        .get("data")
        .and_then(|d| d.get("balance"))
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0);

    let running = container_svc
        .map(|svc| {
            svc.tracked_containers()
                .iter()
                .filter(|c| c.status == ContainerStatus::Running)
                .count()
        })
        .unwrap_or(0);

    let daily_cost = running as f64 * DAILY_COST_PER_CONTAINER;
    let days_remaining = if daily_cost > 0.0 {
        Some(balance / daily_cost)
    } else {
        None
    };

    let mut lines = vec![format!("ElizaCloud credits: ${:.2}", balance)];

    if running > 0 {
        lines.push(format!(
            "Active containers: {} (${:.2}/day) — ~{:.1} days remaining",
            running,
            daily_cost,
            days_remaining.unwrap_or(0.0)
        ));
    } else {
        lines.push("No active containers.".to_string());
    }

    let detailed = options
        .get("detailed")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);

    if detailed {
        let summary_resp = client.get("/credits/summary").await?;
        if let Some(data) = summary_resp.get("data") {
            let total_spent = data["totalSpent"].as_f64().unwrap_or(0.0);
            let total_added = data["totalAdded"].as_f64().unwrap_or(0.0);
            lines.push(format!(
                "Total spent: ${:.2} | Total added: ${:.2}",
                total_spent, total_added
            ));

            if let Some(txns) = data["recentTransactions"].as_array() {
                for tx in txns.iter().take(10) {
                    let amount = tx["amount"].as_f64().unwrap_or(0.0);
                    let sign = if amount >= 0.0 { "+" } else { "" };
                    let desc = tx["description"].as_str().unwrap_or("");
                    let date = tx["created_at"].as_str().unwrap_or("").get(..10).unwrap_or("");
                    lines.push(format!("  {}${:.2} — {} ({})", sign, amount, desc, date));
                }
            }
        }
    }

    let text = lines.join("\n");

    Ok(ActionResult::ok(
        text,
        serde_json::json!({
            "balance": balance,
            "runningContainers": running,
            "dailyCost": daily_cost,
            "estimatedDaysRemaining": days_remaining,
        }),
    ))
}

File: plugins/plugin-elizacloud/actions/check-credits.ts
/**
 * CHECK_CLOUD_CREDITS — Query ElizaCloud credit balance and usage.
 */

import type {
  Action,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import type { CloudAuthService } from "../services/cloud-auth";
import type { CloudContainerService } from "../services/cloud-container";
import type { CreditBalanceResponse, CreditSummaryResponse } from "../types/cloud";

const DAILY_COST_PER_CONTAINER = 0.67;

export const checkCloudCreditsAction: Action = {
  name: "CHECK_CLOUD_CREDITS",
  description: "Check ElizaCloud credit balance, container costs, and estimated remaining runtime.",
  descriptionCompressed: "Check ElizaCloud credits, container costs, remaining runtime.",
  similes: ["check credits", "check balance", "how much credit", "cloud billing"],
  tags: ["cloud", "billing"],
  parameters: [
    {
      name: "detailed",
      description: "Include transaction history",
      required: false,
      schema: { type: "boolean" },
    },
  ],

  validate: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    options?: HandlerOptions
  ): Promise<boolean> => {
    const __avTextRaw = typeof message?.content?.text === "string" ? message.content.text : "";
    const __avText = __avTextRaw.toLowerCase();
    const __avKeywords = ["check", "cloud", "credits"];
    const __avKeywordOk =
      __avKeywords.length > 0 && __avKeywords.some((kw) => kw.length > 0 && __avText.includes(kw));
    const __avRegex = /\b(?:check|cloud|credits)\b/i;
    const __avRegexOk = Boolean(__avText.match(__avRegex));
    const __avSource = String(message?.content?.source ?? "");
    const __avExpectedSource = "";
    const __avSourceOk = __avExpectedSource
      ? __avSource === __avExpectedSource
      : Boolean(__avSource || state || runtime?.agentId || runtime?.getService);
    const __avOptions = options && typeof options === "object" ? options : {};
    const __avInputOk =
      __avText.trim().length > 0 ||
      Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
      Boolean(message?.content && typeof message.content === "object");

    if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
      return false;
    }

    const __avLegacyValidate = async (runtime: IAgentRuntime) => {
      return !!(
        runtime.getService("CLOUD_AUTH") as CloudAuthService | undefined
      )?.isAuthenticated();
    };
    try {
      return Boolean(await __avLegacyValidate(runtime));
    } catch {
      return false;
    }
  },

  async handler(
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    options?: Record<string, unknown>,
    callback?: HandlerCallback
  ): Promise<ActionResult> {
    const auth = runtime.getService("CLOUD_AUTH") as CloudAuthService;
    const containerSvc = runtime.getService("CLOUD_CONTAINER") as CloudContainerService | undefined;
    const client = auth.getClient();

    const detailed =
      options?.detailed === true ||
      (message.metadata as Record<string, unknown> | undefined)?.detailed === true;

    const {
      data: { balance },
    } = await client.get<CreditBalanceResponse>("/credits/balance");

    const running =
      containerSvc?.getTrackedContainers().filter((c) => c.status === "running").length ?? 0;
    const dailyCost = running * DAILY_COST_PER_CONTAINER;
    const daysRemaining = dailyCost > 0 ? balance / dailyCost : null;

    const lines = [
      `ElizaCloud credits: $${balance.toFixed(2)}`,
      running > 0
        ? `Active containers: ${running} ($${dailyCost.toFixed(2)}/day) — ~${daysRemaining?.toFixed(1)} days remaining`
        : "No active containers.",
    ];

    if (detailed) {
      const { data } = await client.get<CreditSummaryResponse>("/credits/summary");
      lines.push(
        `Total spent: $${data.totalSpent.toFixed(2)} | Total added: $${data.totalAdded.toFixed(2)}`
      );
      for (const tx of data.recentTransactions.slice(0, 10)) {
        const sign = tx.amount >= 0 ? "+" : "";
        lines.push(
          `  ${sign}$${tx.amount.toFixed(2)} — ${tx.description} (${new Date(tx.created_at).toLocaleDateString()})`
        );
      }
    }

    const text = lines.join("\n");
    if (callback) await callback({ text, actions: ["CHECK_CLOUD_CREDITS"] });

    return {
      success: true,
      text,
      data: {
        balance,
        runningContainers: running,
        dailyCost,
        estimatedDaysRemaining: daysRemaining,
      },
    };
  },
};
